ogre2/src/Ogre2SegmentationCamera.cc

```diff
@@ -0,0 +1,648 @@
+
+      /// \brief True to generate colored map
+      /// False to generate labels ids map
+      private: bool isColoredMap;
+
+      /// \brief Keep track of num of instances of the same label
+      /// Key: label id, value: num of instances
+      private: std::map<int, int> instancesCount;
+
+      /// \brief keep track of the random colors
+      /// key: encoded key of r,g,b values. value: always True
+      private: std::map<int64_t, bool> takenColors;
+
+      /// \brief keep track of the labels which is already colored
+      /// usful for coloring items in semantic mode in LabelToColor()
+      /// key: label id. value: always True
+      private: std::map<int64_t, bool> coloredLabel;
+
+      private: std::default_random_engine generator;
+
+      // private: std::uniform_int_distribution<int> distribution(1,2);
+
+      /// \brief Ogre2 Scene
+      private: Ogre2ScenePtr scene;
+
+      friend class Ogre2SegmentationCamera;
+    };
+    }
+  }
+}
+
+/////////////////////////////////////////////////
+SegmentationMaterialSwitcher::SegmentationMaterialSwitcher(Ogre2ScenePtr _scene)
+{
+  this->scene = _scene;
+  this->backgroundColor.Set(0, 0, 0);
+
+  // plain material to switch item's material
+  Ogre::ResourcePtr res =
+    Ogre::MaterialManager::getSingleton().load("ign-rendering/plain_color",
+        Ogre::ResourceGroupManager::DEFAULT_RESOURCE_GROUP_NAME);
+
+  this->plainMaterial = res.staticCast<Ogre::Material>();
+  this->plainMaterial->load();
+
+  // plain overlay material
+  this->plainOverlayMaterial =
+      this->plainMaterial->clone("plain_color_overlay");
+  if (!this->plainOverlayMaterial->getTechnique(0) ||
+      !this->plainOverlayMaterial->getTechnique(0)->getPass(0))
+  {
+    ignerr << "Problem creating selection buffer overlay material"
+        << std::endl;
+    return;
+  }
+  Ogre::Pass *overlayPass =
+      this->plainOverlayMaterial->getTechnique(0)->getPass(0);
+  Ogre::HlmsMacroblock macroblock(*overlayPass->getMacroblock());
+  macroblock.mDepthCheck = false;
+  macroblock.mDepthWrite = false;
+  overlayPass->setMacroblock(macroblock);
+}
+
+/////////////////////////////////////////////////
+SegmentationMaterialSwitcher::~SegmentationMaterialSwitcher()
+{
+}
+
+/////////////////////////////////////////////////
+bool SegmentationMaterialSwitcher::IsTakenColor(math::Color _color)
+{
+  // get the int value of the 24 bit color
+  int64_t colorId = _color.R() * 255 * 255 + _color.G() * 255 + _color.B();
+
+  if (this->takenColors.count(colorId))
+    return true;
+  else
+  {
+    this->takenColors[colorId] = true;
+    return false;
+  }
+}
+
+/////////////////////////////////////////////////
+math::Color SegmentationMaterialSwitcher::LabelToColor(int64_t _label)
+{
+  if (_label == this->backgroundLabel)
+    return this->backgroundColor;
+
+  // use label as seed to generate the same color for the label
+  this->generator.seed(_label);
+  std::uniform_int_distribution<int> distribution(0, 255);
```

Comment:
As a future enhancement, it may be helpful to have a configurable colormap so that we can have more deterministic ordering.

Reply:
@mjcarroll do you mean storing the color for each label id of all 255 id ?